modules/core/src/main/java/org/apache/ignite/internal/util/ReentrantReadWriteLockWithTracking.java

```diff
@@ -0,0 +1,330 @@
+
+/** */
+public class ReentrantReadWriteLockWithTracking implements ReadWriteLock {
+    /** Lock hold message. */
+    public static final String LOCK_HOLD_MESSAGE = "ReadLock held the lock more than ";
+
+    /** Lock print threshold. */
+    private long readLockThreshold;
+
+    /** Delegate instance. */
+    private final ReentrantReadWriteLock delegate = new ReentrantReadWriteLock();
+
+    /** Read lock holder. */
+    private ReentrantReadWriteLockWithTracking.ReadLock readLock;
+
+    /** Write lock holder. */
+    private ReentrantReadWriteLockWithTracking.WriteLock writeLock = new ReentrantReadWriteLockWithTracking.WriteLock(delegate);
+
+    /**
+     * ReentrantRWLock wrapper, provides additional trace info on {@link ReadLockWithTracking#unlock()} method, if someone
+     * holds the lock more than {@code readLockThreshold}.
+     *
+     * @param log Ignite logger.
+     * @param readLockThreshold ReadLock threshold timeout.
+     */
+    public ReentrantReadWriteLockWithTracking(IgniteLogger log, long readLockThreshold) {
+        readLock = new ReadLockWithTracking(delegate, log, readLockThreshold);
+
+        this.readLockThreshold = readLockThreshold;
+    }
+
+    /** Delegator implementation. */
+    public ReentrantReadWriteLockWithTracking() {
+        readLock = new ReentrantReadWriteLockWithTracking.ReadLock(delegate);
+    }
+
+    /** {@inheritDoc} */
+    @Override public ReentrantReadWriteLockWithTracking.ReadLock readLock() {
+        return readLock;
+    }
+
+    /** {@inheritDoc} */
+    @Override public ReentrantReadWriteLockWithTracking.WriteLock writeLock() {
+        return writeLock;
+    }
+
+    /** */
+    public long lockWaitThreshold() {
+        return readLockThreshold;
+    }
+
+    /**
+     * Queries if the write lock is held by the current thread.
+     *
+     * @return {@code true} if the current thread holds the write lock and
+     *         {@code false} otherwise
+     */
+    public boolean isWriteLockedByCurrentThread() {
+        return delegate.isWriteLockedByCurrentThread();
+    }
+
+    /**
+     * Queries the number of reentrant read holds on this lock by the
+     * current thread.  A reader thread has a hold on a lock for
+     * each lock action that is not matched by an unlock action.
+     *
+     * @return the number of holds on the read lock by the current thread,
+     *         or zero if the read lock is not held by the current thread
+     */
+    public int getReadHoldCount() {
+        return delegate.getReadHoldCount();
+    }
+
+    /**
+     * Queries the number of read locks held for this lock. This
+     * method is designed for use in monitoring system state, not for
+     * synchronization control.
+     * @return the number of read locks held
+     */
+    public int getReadLockCount() {
+        return delegate.getReadLockCount();
+    }
+
+    /** */
+    public static class WriteLock implements Lock {
```

Comment:
Perhaps this class won't make any sense if we change the return type to java.util.concurrent.locks.ReentrantReadWriteLock.WriteLock in the method above.